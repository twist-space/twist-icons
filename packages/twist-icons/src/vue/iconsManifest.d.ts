export interface IconManifestType {
  id: string;
  name: string;
  author: string;
  projectUrl: string;
  license: string;
  licenseUrl: string;
  total: number;
}
export declare const IconsManifest: IconManifestType[]
