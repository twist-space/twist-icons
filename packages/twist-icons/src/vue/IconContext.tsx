import {
  CSSProperties,
  InjectionKey,
  provide,
  inject,
  SVGAttributes
} from 'vue'

export interface IconContext {
  color?: string
  size?: number
  class?: string
  style?: CSSProperties
  attrs?: SVGAttributes
  styleValue?: string
}

export const IconContextKey = Symbol('IconContextKey') as InjectionKey<IconContext>

export const DefaultContext: IconContext = {
  color: undefined,
  size: undefined,
  class: undefined,
  style: undefined,
  attrs: undefined,
  styleValue: '@twist-space/vue3-icons-css'
}

export const useProvideIconContext = (context: IconContext) => {
  provide(IconContextKey, context)
  return context
}

export const useInjectContext = (IconContextKey) => inject(IconContextKey, DefaultContext)
