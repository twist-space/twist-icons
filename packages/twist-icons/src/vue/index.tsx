export * from './IconBase'
export * from './IconContext'
export * from './useInsertStyles'
