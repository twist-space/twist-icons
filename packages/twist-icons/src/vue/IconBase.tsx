import {
  FunctionalComponent,
  inject,
  PropType,
  SVGAttributes,
  CSSProperties,
  h
} from 'vue'
import { IconContextKey, IconContext, DefaultContext } from './IconContext'
import { UseInsertStyle } from './useInsertStyles'

export interface AbstractNode {
  tag: string
  attrs: Record<string, string>
  children?: AbstractNode[]
}

export interface IconProps extends SVGAttributes {
  size?: number
  color?: string
  style?: CSSProperties
  class?: string
  title?: string
  spin?: boolean
  rotate?: number
}

export type IconType = (props: IconProps) => FunctionalComponent<IconProps>

export function IconBase(props: IconProps, { slots, attrs }) {
  const {
    size,
    title,
    spin,
    rotate,
    ...svgProps
  } = props
  const children = slots.default && slots.default()
  const config = inject<IconContext>(IconContextKey, DefaultContext)
  const mergedSize = size || config.size || '1em'
  let className
  let msTransform = ''
  let transform = ''

  if (config.class) {
    className = config.class
  }
  if (props.class) {
    className = className ? `${className} ${props.class}` : props.class
  }
  if (spin) {
    className ? className += ' twist-icon-loading' : className = 'twist-icon-loading'
  }
  if (rotate) {
    msTransform = `rotate(${rotate}deg)`
    transform = `rotate(${rotate}deg)`
  }

  const titleTag = title ? <title>{title}</title> : ''

  return (
    <svg
      stroke="currentColor"
      fill="currentColor"
      stroke-width="0"
      {...config.attrs}
      {...attrs}
      {...svgProps}
      class={className}
      style={{
        color: props?.color || config.color,
        msTransform,
        transform,
        ...config.style,
        ...props?.style
      }}
      width={mergedSize}
      height={mergedSize}
      xmlns="http://www.w3.org/2000/svg"
    >
      {titleTag}
      {children}
      <UseInsertStyle />
    </svg>
  )
}

IconBase.props = {
  size: Number as PropType<number>,
  color: String as PropType<string>,
  style: Object as PropType<CSSProperties>,
  class: String as PropType<string>,
  title: String as PropType<string>,
  spin: Boolean as PropType<boolean>,
  rotate: Number as PropType<number>,
  onClick: Function as PropType<(e: MouseEvent) => void>
}
IconBase.inheritAttrs = false

export function renderHelper(node: AbstractNode[]) {
  return (
    node &&
    node.map((node, i) =>
      h(node.tag, { key: i, ...node.attrs }, renderHelper(node.children)))
  )
}

export function GenIcon(node: AbstractNode) {
  return (props) => (
    <IconBase {...node.attrs} {...props}>
      {renderHelper(node.children)}
    </IconBase>
  )
}
