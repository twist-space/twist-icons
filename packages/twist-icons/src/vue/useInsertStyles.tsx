import { defineComponent } from 'vue'
import { useInjectContext, IconContextKey } from './IconContext'

const ICONKEY = 'twist-icons-vue3'
const iconCss = `
.twist-icon-loading {
  display: inline-block;
  -webkit-animation: loadingCircle 1s infinite linear;
  animation: twist-icon-circle-loading 1s infinite linear;
}

@-webkit-keyframes twist-icon-circle-loading {
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}
`

function findStyleNode(value: string): HTMLStyleElement | undefined {
  const head = document.querySelector('head') as HTMLHeadElement

  return Array.from(head.children)
    .filter((node) => node.tagName === 'STYLE')
    .find((node) => node.getAttribute(ICONKEY) === value) as HTMLStyleElement | undefined
}

function insertStyles(value: string) {
  const styleTarget = findStyleNode(value)

  if (styleTarget) {
    if (styleTarget.innerHTML !== iconCss) {
      styleTarget.innerHTML = iconCss
    }
    return
  }

  const style = document.createElement('style')
  style.setAttribute('twist-icons-vue3', value)
  style.innerHTML = iconCss
  document.head.appendChild(style)
}

export const InsertStylesComponent = defineComponent({
  name: 'InsertStyles',
  setup() {
    const { styleValue } = useInjectContext(IconContextKey)
    insertStyles(styleValue || '@twist-space/vue3-icons-css')
    return () => null
  }
})
