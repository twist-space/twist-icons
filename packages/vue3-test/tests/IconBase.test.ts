import { mount } from '@vue/test-utils'
import { h } from 'vue'
import { describe, test, expect } from 'vitest'
import { IconProvider } from '../../_vue-icons'
import { AiAudioFilled, AiAudioTwotone } from '../../_vue-icons/ai'

describe('@twist-space/vue3-icons', () => {
  test('Icon component directly render', () => {
    const Icon = mount(AiAudioFilled)
    expect(Icon.html()).toMatchSnapshot()
  })

  test('Two tone Icon render', () => {
    const TwotoneIcon = mount(AiAudioTwotone)
    expect(TwotoneIcon.html()).toMatchSnapshot()
    // Two tone icon path must than one path
    expect(TwotoneIcon.findAll('path')).not.toHaveLength(1)
  })

  test('Icon component props <size, color, title>', () => {
    const Icon = mount(AiAudioFilled, {
      props: {
        size: 20,
        color: 'red',
        title: 'twist-vue3-icon'
      }
    })

    expect(Icon.html()).toMatchSnapshot()
    expect(Icon.html()).toContain('style="color: red;"')
    expect(Icon.html()).toContain('width="20" height="20"')
    expect(Icon.html()).toContain('<title>twist-vue3-icon</title>')
  })

  test('Icon component props <spin>', () => {
    const Icon = mount(AiAudioFilled, {
      props: {
        spin: true
      }
    })

    expect(Icon.html()).toMatchSnapshot()
    expect(Icon.classes('twist-icon-loading')).toBe(true)
  })

  test('Icon component props <rotate>', () => {
    const Icon = mount(AiAudioFilled, {
      props: {
        rotate: 60
      }
    })

    expect(Icon.html()).toMatchSnapshot()
    expect(Icon.html()).toContain('style="transform: rotate(60deg);"')
  })
})

describe('IconConfigProvider and Icon component render priority test', () => {
  test('Icon props hsa a higher priority than IconConfigProvider props', () => {
    const AiAudioFilledIcon = h(AiAudioFilled, { size: 200, color: 'green' })
    const Icon = mount(IconProvider, {
      props: {
        size: 100,
        color: 'black'
      },
      slots: {
        default: AiAudioFilledIcon
      }
    })

    expect(Icon.html()).toMatchSnapshot()
    expect(Icon.html()).toContain('width="200" height="200"')
    expect(Icon.html()).toContain('style="color: green;"')
  })
})
