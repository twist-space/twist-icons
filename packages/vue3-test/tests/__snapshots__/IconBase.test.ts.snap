// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@twist-space/vue3-icons > Icon component directly render 1`] = `
"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024" width="1em" height="1em">
  <path fill="currentColor" d="M512 624c93.9 0 170-75.2 170-168V232c0-92.8-76.1-168-170-168s-170 75.2-170 168v224c0 92.8 76.1 168 170 168m330-170c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8c0 140.3-113.7 254-254 254S258 594.3 258 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8c0 168.7 126.6 307.9 290 327.6V884H326.7c-13.7 0-24.7 14.3-24.7 32v36c0 4.4 2.8 8 6.2 8h407.6c3.4 0 6.2-3.6 6.2-8v-36c0-17.7-11-32-24.7-32H548V782.1c165.3-18 294-158 294-328.1"></path>
  <!---->
</svg>"
`;

exports[`@twist-space/vue3-icons > Icon component props <rotate> 1`] = `
"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024" style="transform: rotate(60deg);" width="1em" height="1em">
  <path fill="currentColor" d="M512 624c93.9 0 170-75.2 170-168V232c0-92.8-76.1-168-170-168s-170 75.2-170 168v224c0 92.8 76.1 168 170 168m330-170c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8c0 140.3-113.7 254-254 254S258 594.3 258 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8c0 168.7 126.6 307.9 290 327.6V884H326.7c-13.7 0-24.7 14.3-24.7 32v36c0 4.4 2.8 8 6.2 8h407.6c3.4 0 6.2-3.6 6.2-8v-36c0-17.7-11-32-24.7-32H548V782.1c165.3-18 294-158 294-328.1"></path>
  <!---->
</svg>"
`;

exports[`@twist-space/vue3-icons > Icon component props <size, color, title> 1`] = `
"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024" style="color: red;" width="20" height="20">
  <title>twist-vue3-icon</title>
  <path fill="currentColor" d="M512 624c93.9 0 170-75.2 170-168V232c0-92.8-76.1-168-170-168s-170 75.2-170 168v224c0 92.8 76.1 168 170 168m330-170c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8c0 140.3-113.7 254-254 254S258 594.3 258 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8c0 168.7 126.6 307.9 290 327.6V884H326.7c-13.7 0-24.7 14.3-24.7 32v36c0 4.4 2.8 8 6.2 8h407.6c3.4 0 6.2-3.6 6.2-8v-36c0-17.7-11-32-24.7-32H548V782.1c165.3-18 294-158 294-328.1"></path>
  <!---->
</svg>"
`;

exports[`@twist-space/vue3-icons > Icon component props <spin> 1`] = `
"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024" class="twist-icon-loading" width="1em" height="1em">
  <path fill="currentColor" d="M512 624c93.9 0 170-75.2 170-168V232c0-92.8-76.1-168-170-168s-170 75.2-170 168v224c0 92.8 76.1 168 170 168m330-170c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8c0 140.3-113.7 254-254 254S258 594.3 258 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8c0 168.7 126.6 307.9 290 327.6V884H326.7c-13.7 0-24.7 14.3-24.7 32v36c0 4.4 2.8 8 6.2 8h407.6c3.4 0 6.2-3.6 6.2-8v-36c0-17.7-11-32-24.7-32H548V782.1c165.3-18 294-158 294-328.1"></path>
  <!---->
</svg>"
`;

exports[`@twist-space/vue3-icons > Two tone Icon render 1`] = `
"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024" width="1em" height="1em">
  <path fill="currentColor" fill-opacity=".15" d="M512 552c54.3 0 98-43.2 98-96V232c0-52.8-43.7-96-98-96s-98 43.2-98 96v224c0 52.8 43.7 96 98 96"></path>
  <path fill="currentColor" d="M842 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8c0 140.3-113.7 254-254 254S258 594.3 258 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8c0 168.7 126.6 307.9 290 327.6V884H326.7c-13.7 0-24.7 14.3-24.7 32v36c0 4.4 2.8 8 6.2 8h407.6c3.4 0 6.2-3.6 6.2-8v-36c0-17.7-11-32-24.7-32H548V782.1c165.3-18 294-158 294-328.1"></path>
  <path fill="currentColor" d="M512 624c93.9 0 170-75.2 170-168V232c0-92.8-76.1-168-170-168s-170 75.2-170 168v224c0 92.8 76.1 168 170 168m-98-392c0-52.8 43.7-96 98-96s98 43.2 98 96v224c0 52.8-43.7 96-98 96s-98-43.2-98-96z"></path>
  <!---->
</svg>"
`;

exports[`IconConfigProvider and Icon component render priority test > Icon props hsa a higher priority than IconConfigProvider props 1`] = `
"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024" style="color: green;" width="200" height="200">
  <path fill="currentColor" d="M512 624c93.9 0 170-75.2 170-168V232c0-92.8-76.1-168-170-168s-170 75.2-170 168v224c0 92.8 76.1 168 170 168m330-170c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8c0 140.3-113.7 254-254 254S258 594.3 258 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8c0 168.7 126.6 307.9 290 327.6V884H326.7c-13.7 0-24.7 14.3-24.7 32v36c0 4.4 2.8 8 6.2 8h407.6c3.4 0 6.2-3.6 6.2-8v-36c0-17.7-11-32-24.7-32H548V782.1c165.3-18 294-158 294-328.1"></path>
  <!---->
</svg>"
`;
